src/systems/filecoin_markets/storage_market/storage_market_actor_state.go

```diff
@@ -0,0 +1,288 @@
+// This is the check before a StorageDeal appears onchain
+// It checks the following:
+//   - verify deal did not expire when it is signed
+//   - verify deal hits the chain before StartEpoch
+//   - verify client and provider address and signature are correct (TODO may not be needed)
+//   - verify StorageDealCollateral match requirements for MinimumStorageDealCollateral
+//   - verify client and provider has sufficient balance
+func (st *StorageMarketActorState_I) _validateNewStorageDeal(rt Runtime, d deal.StorageDeal) bool {
+
+	p := d.Proposal()
+
+	st._assertValidClienSignature(rt, p)
+	st._assertValidDealTimingAtPublish(rt, p)
+	st._assertValidDealMinimum(rt, p)
+	st._assertSufficientBalanceAvailForDeal(rt, p)
+
+	return true
+}
+
+func (st *StorageMarketActorState_I) _activateDeal(rt Runtime, deal deal.OnchainDeal) deal.OnchainDeal {
+
+	dealP := deal.Deal().Proposal()
+	deal.Impl().LastPaymentEpoch_ = dealP.StartEpoch()
+	st.Deals()[deal.ID()] = deal
+
+	return deal
+}
+
+// TODO: consider returning a boolean
+func (st *StorageMarketActorState_I) _lockBalance(rt Runtime, addr addr.Address, amount actor.TokenAmount) {
+	if amount < 0 {
+		rt.Abort("sma._lockBalance: negative amount.")
+	}
+
+	currBalance, found := st.Balances()[addr]
+	if !found {
+		rt.Abort("sma._lockBalance: addr not found.")
+	}
+
+	currBalance.Impl().Available_ -= amount
+	currBalance.Impl().Locked_ += amount
+}
+
+func (st *StorageMarketActorState_I) _unlockBalance(rt Runtime, addr addr.Address, amount actor.TokenAmount) {
+	if amount < 0 {
+		rt.Abort("sma._unlockBalance: negative amount.")
+	}
+
+	currBalance, found := st.Balances()[addr]
+	if !found {
+		rt.Abort("sma._unlockBalance: addr not found.")
+	}
+
+	currBalance.Impl().Locked_ -= amount
+	currBalance.Impl().Available_ += amount
+}
+
+// move funds from locked in client to available in provider
+func (st *StorageMarketActorState_I) _transferBalance(rt Runtime, fromLocked addr.Address, toAvailable addr.Address, amount actor.TokenAmount) {
+	fromB := st.Balances()[fromLocked]
+	toB := st.Balances()[toAvailable]
+
+	if fromB.Locked() < amount {
+		rt.Abort("sma._transferBalance: attempt to unlock funds greater than actor has")
+		return
+	}
+
+	fromB.Impl().Locked_ -= amount
+	toB.Impl().Available_ += amount
+}
+
+func (st *StorageMarketActorState_I) _lockFundsForStorageDeal(rt Runtime, deal deal.StorageDeal) {
+	p := deal.Proposal()
+
+	st._lockBalance(rt, p.Client(), p.ClientBalanceRequirement())
+	st._lockBalance(rt, p.Provider(), p.ProviderBalanceRequirement())
+}
+
+func (st *StorageMarketActorState_I) _getOnchainDeal(rt Runtime, dealID deal.DealID) deal.OnchainDeal {
+	deal, found := st.Deals()[dealID]
+	if !found {
+		rt.Abort("sm._getOnchainDeal: dealID not found in Deals.")
+	}
+
+	return deal
+}
+
+func (st *StorageMarketActorState_I) _assertPublishedDealState(rt Runtime, dealID deal.DealID) {
+
+	deal := st._getOnchainDeal(rt, dealID)
```

Comment:
abort message will be unclear in such cases. consider changing _getOnChainDeal to return both a deal and status (bool) so you can abort in caller method for better logging

Reply:
as with balances

Reply:
renaming to _safeGetOnChainDeal and _safeGetBalance